### PR TITLE
[ADD] Add section Time spent tasks

### PR DESCRIPTION
**Pull request by jananjoy**

### CODE REVIEW

File: project_event/views/project_task_view.xml [hunk 1]
```diff
@@ -85,6 +85,20 @@
                         </page>
                         <page string="Technical Sheet">
                             <field name="spectators"/>
+                            <group>
+                                <page string="Time Spent">
+                                    <field name="child_ids">
```
Comment (rimbendhaou) [inline project_event/views/project_task_view.xml:89]: we dont need a page if we already have a page named technical sheet. Display the name of the section Time spent.

File: project_event/models/project_task.py [hunk 2]
```diff
@@ -216,6 +217,33 @@ def check_numeric(self):
         except Exception as exception:
             raise ValidationError(_('The Input value should be numeric'))
 
+    real_date_start = fields.Datetime(
+        string='Actual Start Time',
+        default=None,
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:222]: by default it should be the same date as date_start

Comment (rimbendhaou) [inline project_event/models/project_task.py:222]: by default it should be the same date as date_start

File: project_event/models/project_task.py [hunk 3]
```diff
@@ -216,6 +217,33 @@ def check_numeric(self):
...
+        index=True,
+        copy=False,
+        track_visibility='onchange',
+    )
+    real_date_end = fields.Datetime(
+        string='Actual End Time',
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:225]: by default it should be the same date as date_end

File: project_event/models/project_task.py [hunk 4]
```diff
@@ -216,6 +217,33 @@ def check_numeric(self):
         except Exception as exception:
             raise ValidationError(_('The Input value should be numeric'))
 
+    real_date_start = fields.Datetime(
+        string='Actual Start Time',
+        default=None,
+        index=True,
+        copy=False,
+        track_visibility='onchange',
```
Comment (rimbendhaou) [inline project_event/models/project_task.py]: no need to add check visibility, no index or copy

File: project_event/models/project_task.py [hunk 5]
```diff
@@ -216,6 +217,33 @@ def check_numeric(self):
...
+    )
+    real_date_end = fields.Datetime(
+        string='Actual End Time',
+        index=True,
```
Comment (rimbendhaou) [inline project_event/models/project_task.py]: no need for index or copy

File: project_event/models/project_task.py [hunk 6]
```diff
@@ -216,6 +217,33 @@ def check_numeric(self):
...
+    total_time = fields.Char(
+        string='Actual Total Time',
+        compute='_compute_actual_total_time',
+    )
+
+    @api.one
+    @api.depends('real_date_start', 'real_date_end')
+    def _compute_actual_total_time(self):
+        self.ensure_one()
+        time_diff = relativedelta(fields.Datetime.from_string(self.real_date_end),
```
Comment (rimbendhaou) [inline project_event/models/project_task.py]: what if theses two fields are not mentioned

Comment (rimbendhaou) [reply project_event/models/project_task.py]: you should add a test for this new functionality

**Review by rimbendhaou — CHANGES_REQUESTED** [review]

File: project_event/models/project_task.py [hunk 7]
```diff
@@ -216,6 +217,34 @@ def check_numeric(self):
         except Exception as exception:
             raise ValidationError(_('The Input value should be numeric'))
 
+    real_date_start = fields.Datetime(
+        string='Actual Start Time',
+        default=None,
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:222]: @jananjoy what about default? by default it should be the same date as date_start

File: project_event/models/project_task.py [hunk 8]
```diff
@@ -216,6 +217,34 @@ def check_numeric(self):
...
+    )
+    real_date_end = fields.Datetime(
+        string='Actual End Time',
+        default=None,
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:226]: by default it should be the same date as date_end

File: project_event/tests/common.py [hunk 9]
```diff
@@ -163,3 +163,15 @@ def setUp(self):
...
+            'name': 'Sample task1',
+            'activity_task_type': 'activity',
+            'partner_id': self.project_1.partner_id.id,
+            'room_id': self.room_1.id,
+            'date_start': fields.Datetime.to_string(datetime.today()),
+            'date_end': fields.Datetime.to_string(datetime.today() +
+                                                  timedelta(hours=4)),
+            'real_date_start': fields.Datetime.to_string(datetime.today()),
+            'real_date_end': fields.Datetime.to_string(datetime.today()),
+            'total_time': "1:1",
```
Comment (rimbendhaou) [inline project_event/tests/common.py]: This field should be filled automatically

File: project_event/tests/test_project_event_task.py [hunk 10]
```diff
@@ -429,3 +429,8 @@ def test_160_check_partner_id_calendar_event_as_client_of_task(self):
             calendar_event.client_id.id,
             self.partner_1.id
         )
+
+    def test_170_compute_actual_total_time(self):
+        self.assertEqual(self.task_3.name, 'Sample task1')
+        diff = self.task_3.total_time
+        self.assertEqual(diff, "0:0")
```
Comment (rimbendhaou) [inline project_event/tests/test_project_event_task.py:435]: better to test with a time difference>0

File: project_event/tests/common.py [hunk 11]
```diff
@@ -163,3 +163,15 @@ def setUp(self):
...
+        self.task_3 = self.Tasks.create({
+            'name': 'Sample task1',
+            'activity_task_type': 'activity',
+            'partner_id': self.project_1.partner_id.id,
+            'room_id': self.room_1.id,
+            'date_start': fields.Datetime.to_string(datetime.today()),
+            'date_end': fields.Datetime.to_string(datetime.today() +
+                                                  timedelta(hours=4)),
+            'real_date_start': fields.Datetime.to_string(datetime.today()),
+            'real_date_end': fields.Datetime.to_string(datetime.today()),
```
Comment (rimbendhaou) [inline project_event/tests/common.py:175]: better to have real_date_end different then real_date_start

File: project_event/models/project_task.py [hunk 12]
```diff
@@ -216,6 +217,34 @@ def check_numeric(self):
...
+    )
+    total_time = fields.Char(
+        string='Actual Total Time',
+        compute='_compute_actual_total_time',
+    )
+
+    @api.one
```
Comment (rimbendhaou) [inline project_event/models/project_task.py:233]: You have to add also a constraint that checks if real_date_end > real_date_start

**Review by rimbendhaou — CHANGES_REQUESTED** [review]

File: project_event/models/project_task.py [hunk 13]
```diff
@@ -216,6 +217,41 @@ def check_numeric(self):
...
+                time_diff = relativedelta(fields.Datetime.from_string(self.real_date_end),
+                                          fields.Datetime.from_string(self.real_date_start))
+                hours = time_diff.hours
+                minutes = time_diff.minutes
+                self.total_time = str(hours) + ":" + str(minutes)
+            elif self.real_date_end == self.real_date_start:
+                self.total_time = "0:0"
+            else:
+                raise ValidationError(_('Actual Start Time should be before Actual End Time'))
+        elif self.real_date_start is None:
```
Comment (eilst) [inline project_event/models/project_task.py]: This forces the real time fields to be filled, for now we need those fields empty by default

**Review by eilst — APPROVED** [review]

**Pull request merged**